source/sliding_puzzle.py

```diff
@@ -124,7 +139,67 @@ def _join(self, row: List[PuzzlePiece]) -> List[str]:
     def draw(self) -> str:
         """Draw the puzzle in its current state."""
         output = []
-        for row in self.pieces:
+        for row in self.rows:
             output.extend(self._join(row))
 
         return '\n'.join(output)
+
+    def shuffle(self) -> None:
+        """Shuffle the puzzle by randomizing the order of the pieces"""
+        for row in self.rows:
+            random.shuffle(row)
+        random.shuffle(self.rows)
+        # empty one of the rows
+        random.choice(random.choice(self.rows)).clear()
+
+    def _get_empty_piece_position(self) -> PiecePosition:
+        x, y, index = [
+            (x, y, piece.index)
+            for x, row in enumerate(self.rows)
+            for y, piece in enumerate(row)
+            if piece._data == []
```

Comment:
```suggestion
            if piece.empty
```

Reply:
I try the code with empty, it doesn't work, try it yourself, all non empty pieces are evaluated to empty == True, and the empty value of the empty pieces is False, I didn't know why, so I ignore it

Reply:
Ahhh.. yeah I see the bug now.